packages/@aws-cdk/assertions/lib/match.ts

```diff
@@ -1,13 +1,242 @@
+    super();
+    this.partialObjects = options.partialObjects ?? false;
+
+    if (Match.isMatcher(this.pattern)) {
+      throw new Error('ExactMatch cannot be nested with another matcher at the top level. Deeper nesting is allowed.');
+    }
+  }
+
+  public test(actual: any): MatchFailure[] {
+    if (Array.isArray(this.pattern)) {
+      return new ArrayMatch(this.pattern, { partial: false }).test(actual);
+    }
+
+    if (typeof this.pattern === 'object') {
+      return new ObjectMatch(this.pattern, { partial: this.partialObjects }).test(actual);
+    }
+
+    if (typeof this.pattern !== typeof actual) {
+      return [{ path: [], message: `Expected type ${typeof this.pattern} but received ${getType(actual)}` }];
+    }
+
+    if (this.pattern === ABSENT) {
+      throw new Error('ABSENT can only be used in an object matcher');
+    }
+
+    if (actual !== this.pattern) {
+      return [{ path: [], message: `Expected ${this.pattern} but received ${actual}` }];
+    }
+
+    return [];
+  }
+}
+
+/**
+ * Options when initializing the `ArrayMatch` class.
+ */
+export interface ArrayMatchOptions {
+  /**
+   * Whether the pattern should partially match with the target.
+   * The target array can contain more elements than expected by the pattern.
+   * Matching elements in the target must be present in the same order as in the pattern.
+   * @default true
+   */
+  readonly partial?: boolean;
+}
+
+/**
+ * Match class that matches arrays.
+ */
+export class ArrayMatch extends Match {
+  private readonly partial: boolean;
+
+  constructor(private readonly pattern: any[], options: ArrayMatchOptions = {}) {
+    super();
+    this.partial = options.partial ?? true;
+  }
+
+  public test(actual: any): MatchFailure[] {
+    if (!Array.isArray(actual)) {
+      return [{ path: [], message: `Expected type array but received ${getType(actual)}` }];
+    }
+    if (!this.partial && this.pattern.length !== actual.length) {
+      return [{ path: [], message: `Expected array of length ${this.pattern.length} but received ${actual.length}` }];
+    }
+
+    let patternIdx = 0;
+    let actualIdx = 0;
+
+    const failures: MatchFailure[] = [];
+    while (patternIdx < this.pattern.length && actualIdx < actual.length) {
+      let patternElement = this.pattern[patternIdx];
+      let matcher = Match.isMatcher(patternElement) ? patternElement : new LiteralMatch(patternElement);
+      const innerFailures = matcher.test(actual[actualIdx]);
+
+      if (!this.partial || innerFailures.length === 0) {
+        failures.push(...composeFailures(`[${actualIdx}]`, innerFailures));
+        patternIdx++;
+        actualIdx++;
+      } else {
+        actualIdx++;
+      }
+    }
+
+    for (; patternIdx < this.pattern.length; patternIdx++) {
+      const p = this.pattern[patternIdx];
+      const e = (Match.isMatcher(p) || typeof p === 'object') ? ' ' : ` [${p}] `;
+      failures.push({ path: [], message: `Missing element${e}at pattern index ${patternIdx}` });
```

Comment:
nitpick (the square brackets kinda make it look like a singleton array is expected)
```suggestion
      const pattern = this.pattern[patternIdx];
      const element = (Match.isMatcher(pattern) || typeof pattern === 'object') ? '' : `${pattern} `;
      failures.push({ path: [], message: `Missing element ${element}at pattern index ${patternIdx}` });
```